### PR TITLE
Feat: lsp support find refs

### PR DESCRIPTION
**Pull request by amyXia1994**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [X] Y 
re #712 

#### 2. What is the scope of this PR (e.g. component or file name):

added: implement the find reference API and add some integration tests

kclvm/tools/src/LSP/src/find_refs.rs
kclvm/tools/src/LSP/src/test_data/find_refs_test/...

modified: 
- add common logic of build word index
- add find reference related lsp provider register, handler
- fix the `kcl_pos` bug in `from_lsp.rs`
- update the `Definition` enum, add a name field in `kclvm/tools/src/LSP/src/goto_def.rs` and update relative usages of it

kclvm/tools/src/LSP/src/capabilities.rs
kclvm/tools/src/LSP/src/completion.rs
kclvm/tools/src/LSP/src/from_lsp.rs
kclvm/tools/src/LSP/src/goto_def.rs
kclvm/tools/src/LSP/src/hover.rs
kclvm/tools/src/LSP/src/lib.rs
kclvm/tools/src/LSP/src/main.rs
kclvm/tools/src/LSP/src/request.rs
kclvm/tools/src/LSP/src/state.rs
kclvm/tools/src/LSP/src/tests.rs
kclvm/tools/src/LSP/src/util.rs

deleted: the older incomplete implementation attempt about find ref
kclvm/tools/src/LSP/src/find_ref/...

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):
Feat: lsp support basic find reference

Known flaws:
- the performance in a large-scaled workspace is slow

- the `scope` to find references needs to be narrowed down to min(workspace, kcl_package_root). now the find refs searches under the kcl package root(identified by the kcl.mod file)
- does not handle the include_declaration param. now always returns refs include declaration
- the references of an imported package is not supported yet.

- [X] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [X] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [X] Unit test
- [X] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

File: kclvm/tools/src/LSP/src/util.rs [hunk 1]
```diff
@@ -737,3 +738,411 @@ pub(crate) fn get_pkg_scope(
...
+
+/// scan and build a word -> Locations index map
+pub fn build_word_index(path: String) -> anyhow::Result<HashMap<String, Vec<Location>>> {
+    let mut index: HashMap<String, Vec<Location>> = HashMap::new();
+    if let Ok(files) = get_kcl_files(path.clone(), true) {
+        for file_path in &files {
+            // str path to url
+            if let Ok(url) = Url::from_file_path(file_path) {
+                // read file content and save the word to word index
+                let text = read_file(file_path).unwrap();
```
Comment (He1pa) [inline kclvm/tools/src/LSP/src/util.rs:750]: let text = read_file(file_path)?;

File: kclvm/tools/src/LSP/src/goto_def.rs [hunk 2]
```diff
@@ -68,18 +68,18 @@ pub(crate) fn goto_definition(
 
 #[derive(Debug)]
 pub enum Definition {
-    Object(ScopeObject),
-    Scope(Scope),
+    Object(ScopeObject, String),
```
[the remark this answers is not in the document]
Comment (amyXia1994) [reply kclvm/tools/src/LSP/src/goto_def.rs:71]: I thinked about using `ScopeObject.name` but when the definition matches a Scope(such as a schema definition), there's no name for the scope.

File: kclvm/tools/src/LSP/src/util.rs (hunk 1 above)
Comment (amyXia1994) [reply kclvm/tools/src/LSP/src/util.rs]: fixed

File: kclvm/tools/src/LSP/src/request.rs [hunk 3]
```diff
@@ -42,7 +44,9 @@ impl LanguageServerState {
                 state.shutdown_requested = true;
                 Ok(())
             })?
+            // .on::<lsp_types::request::Initialize>(handle_initialize)?
```
[the remark this answers is not in the document]
Comment (amyXia1994) [reply kclvm/tools/src/LSP/src/request.rs]: fixed

File: kclvm/tools/src/LSP/src/util.rs [hunk 4]
```diff
@@ -737,3 +738,411 @@ pub(crate) fn get_pkg_scope(
...
+    let mut prev_word = false;
+    let mut words: Vec<Word> = vec![];
+    for (i, ch) in chars.iter().enumerate() {
+        let is_id_start = rustc_lexer::is_id_start(*ch);
+        let is_id_continue = rustc_lexer::is_id_continue(*ch);
+        // If the character is valid identfier start and the previous character is not valid identifier continue, mark the start position.
+        if is_id_start && !prev_word {
+            start_pos = i;
+        }
+        match is_id_continue {
```
Comment (NeverRaR) [inline kclvm/tools/src/LSP/src/util.rs:813]: replace
```
match is_id_continue {
    ...
}
```
with 
```
if is_id_continue {
    ...
} else {
   ...
}

```

File: kclvm/tools/src/LSP/src/find_refs.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    cursor_path: String,
+    logger: F,
+) -> anyhow::Result<Option<Vec<Location>>> {
+    // todo: decide the scope by the workspace root and the kcl.mod both, use the narrower scope
+    // todo: should use the current file path
+    if let Some(root) = get_pkg_root(def_loc.uri.path()) {
+        match build_word_index(root) {
+            std::result::Result::Ok(word_index) => {
+                if let Some(locs) = word_index.get(name.as_str()).cloned() {
+                    return anyhow::Ok(Some(
```
Comment (Peefy) [inline kclvm/tools/src/LSP/src/find_refs.rs:20]: Remove the `return` word.

File: kclvm/tools/src/LSP/src/find_refs.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+                                match parse_param_and_compile(
+                                    Param {
+                                        file: file_path.clone(),
+                                    },
+                                    None,
+                                ) {
+                                    Ok((prog, scope, _)) => {
+                                        let ref_pos = kcl_pos(&file_path, ref_loc.range.start);
+                                        // find def from the ref_pos
+                                        if let Some(real_def) =
```
Comment (Peefy) [inline kclvm/tools/src/LSP/src/find_refs.rs:35]: U can use the `cargo clippy` or the `cargo fix` command to simply all match statements.

File: kclvm/tools/src/LSP/src/util.rs (hunk 4 above)
Comment (amyXia1994) [reply kclvm/tools/src/LSP/src/util.rs]: thanks, fixed

File: kclvm/tools/src/LSP/src/find_refs.rs (hunk 5 above)
Comment (amyXia1994) [reply kclvm/tools/src/LSP/src/find_refs.rs]: fixed

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**